lms/resources/lti_launch.py

```diff
@@ -2,15 +2,23 @@
 import functools
 import hashlib
 
-from pyramid.httpexceptions import HTTPBadRequest
 from pyramid.security import Allow
 
 from lms.resources._js_config import JSConfig
 from lms.values import HUser
 
 
 class LTILaunchResource:
-    """Context resource for LTI launch requests."""
+    """
+    Context resource for LTI launch requests.
+
+    Many methods and properties of this class are only meant to be called when
+    request.parsed_params holds validated params from an LTI launch request and
+    might crash otherwise. So you should only call these methods after the
+    request has been validated with BasicLTILaunchSchema or similar (for
+    example from views that have schema=BasicLTILaunchSchema in their view
+    config).
+    """
```

Comment:
Yeah, there's no real way. In general I think it's acceptable for a method to say "You must only call me when X" in its docstring.

It's possible that context resources shouldn't be reading a whole lot of request params like `LTILaunchResource` does. If a method that a context resource wants to provide requires a request param, the method should have that param as an argument, rather than reading it from the request itself.

We might end up moving bit by bit in that direction with `LTILaunchResource` but don't want to bite it all off at once 'cos there's a lot here and I'm not 100% sure about the direction yet